src/main/java/org/opensearch/neuralsearch/query/HybridQuery.java

```diff
@@ -106,22 +112,40 @@ public Query rewrite(IndexSearcher indexSearcher) throws IOException {
             return new MatchNoDocsQuery("empty HybridQuery");
         }
 
-        boolean actuallyRewritten = false;
-        List<Query> rewrittenSubQueries = new ArrayList<>();
+        final HybridQueryRewriteCollectorManager manager = new HybridQueryRewriteCollectorManager(indexSearcher);
+        final List<Callable<Void>> queryRewriteTasks = new ArrayList<>();
+        final List<HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>>> collectors = new ArrayList<>();
         for (Query subQuery : subQueries) {
-            Query rewrittenSub = subQuery.rewrite(indexSearcher);
-            /* we keep rewrite sub-query unless it's not equal to itself, it may take multiple levels of recursive calls
-               queries need to be rewritten from high-level clauses into lower-level clauses because low-level clauses
-               perform better. For hybrid query we need to track progress of re-write for all sub-queries */
-            actuallyRewritten |= rewrittenSub != subQuery;
-            rewrittenSubQueries.add(rewrittenSub);
+            final HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>> collector = manager.newCollector();
+            collectors.add(collector);
+            queryRewriteTasks.add(() -> rewriteQuery(subQuery, collector));
         }
 
-        if (actuallyRewritten) {
-            return new HybridQuery(rewrittenSubQueries);
+        HybridQueryExecutor.getExecutor().invokeAll(queryRewriteTasks);
+
+        final boolean isAnyQueryRewritten = manager.anyQueryRewrite(collectors);
+        if (isAnyQueryRewritten == false) {
+            return super.rewrite(indexSearcher);
         }
+        final List<Query> rewrittenSubQueries = manager.getQueriesAfterRewrite(collectors);
+        return new HybridQuery(rewrittenSubQueries);
+    }
 
-        return super.rewrite(indexSearcher);
+    private Void rewriteQuery(Query query, HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>> collector) {
+        collector.collect(indexSearcher -> {
+            try {
+                Query rewrittenQuery = query.rewrite(indexSearcher);
+                /* we keep rewrite sub-query unless it's not equal to itself, it may take multiple levels of recursive calls
+                queries need to be rewritten from high-level clauses into lower-level clauses because low-level clauses
+                perform better. For hybrid query we need to track progress of re-write for all sub-queries */
+
+                boolean actuallyRewritten = rewrittenQuery != query;
+                return new AbstractMap.SimpleEntry(rewrittenQuery, actuallyRewritten);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```

Comment:
It seems better to use IllegalStateException since it indicates this is an server internal exception, also should add error messages.